### ISSUE TITLE
Completion of error handling

### ISSUE THREAD

**Issue by elfring** [RobertElderSoftware/recc#4]
Would you like to add more error handling for return values from functions like the following?
- [fprintf](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fprintf.html "Print formatted output.") ⇒ [output_start_end](https://github.com/RobertElderSoftware/recc/blob/9389ef35121df604898f2f00c43a5bbd97a08afa/preloader.c#L433)
- [malloc](http://pubs.opengroup.org/onlinepubs/9699919799/functions/malloc.html "Memory allocation") ⇒ [memory_pool_collection_create](https://github.com/RobertElderSoftware/recc/blob/9389ef35121df604898f2f00c43a5bbd97a08afa/memory_pool_collection.c#L18)

**Comment by RobertElder** [comment]
Yeah, the error handling right now is pretty much non existent.  Once I get close to having the compiler work well for all of basic C89, and have some very detailed unit tests, then I'll starting adding some better error handling.  For now instant and catastrophic failure using assertions works pretty well.

**Comment by elfring** [comment]
I suggest to avoid [ignorance of return values](https://cwe.mitre.org/data/definitions/252.html "Unchecked return values") a bit more.

Are you interested to apply [aspect-oriented software development](http://aspectc.org/ "AspectC++")?
How do you think about to encapsulate error detection and corresponding exception handling as [a reusable aspect](http://en.wikipedia.org/wiki/Aspect_%28computer_programming%29 "Combination of a pointcut with advice") in your software?

**Comment by RobertElder** [comment]
Error recovery and handling is a complex topic, especially in a language like C, where you don't have exceptions.  I have a few ideas for how I'll implement error handling, but for now I'm just going to rely on instant (and obvious) program failure when an error is encountered.  Once the compiler is able to work with some non-trivial test cases and I can fix it to some good unit tests, I'll start refactoring the error handling to be more user friendly.  Currently, it is expected that no error will occur during compillation, so it would always be best to test your program first with gcc or clang with full warnings.  I'm not following any particular development processes other than following my own personal opinions and experience on software development.

**Comment by elfring** [comment]
How do you think about to improve [static source code analysis](https://en.wikipedia.org/wiki/List_of_tools_for_static_code_analysis#C.2FC.2B.2B "Tools for static code analysis") also for your software?

Do you find information sources like the following useful?
* Article "[A Reusable Aspect for Memory Allocation Checking](http://aspectc.net/ac_tutorial.html#reusablemachecking "AspectC Tutorial")"
* Document "[Aspect-Oriented Software Development - An Introduction](http://www.info.ucl.ac.be/~jbrichau/courses/introductionToAOSD.pdf "AOSD")" by Johan Brichau and Theo D'Hondt
* Paper "[On Typesafe Aspect Implementations in C++](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.125.2164 "AspectC++")" by Daniel Lohmann and Olaf Spinczyk

**Comment by RobertElder** [comment]
Yeah, static analysis is something that I think about a lot.  In the future, I'd like to build some of my own static analysis tools into the compiler.  Currently, I only rely on static analysis that comes from the warning flags from gcc and clang.
